### ISSUE TITLE
Support for password and JWT/SAML SSO grant types in OAuthMiddleware

### ISSUE THREAD

**Issue by bgever** [aspnet/Security#1236]
Related to #1036, I'd like to see support for:

- [Resource Owner Password Credentials](https://tools.ietf.org/html/rfc6749#section-1.3.3) (password) authorization grant.
- JSON Web Token (JWT) Profile for OAuth 2.0 Client Authentication and Authorization Grants ([RFC 7523](https://tools.ietf.org/html/rfc7523))
- Security Assertion Markup Language (SAML) 2.0 Profile for OAuth 2.0 Client Authentication and Authorization Grants ([RFC 7522](https://tools.ietf.org/html/rfc7522))

The current OAuthMiddleware seems to support the **Authorisation Code** grant and its redirect flow *only*. (It would be nice to make this more explicit in the documentation)

All three above mentioned grant types are critical for our login needs, and I've currently built custom support for these in our ASP.NET 4.6 app. As I'm refactoring the codebase to ASP.NET Core, I'm now in need to have this support, or build a custom implementation.

If there are no immediate plans for this, I'm happy to entertain the idea of making a PR to add this support to the OAuthMiddleware. It will likely need some design discussions, and I may need guidance on the approach for this PR so it can be merged in.

Interested to hear your thoughts @HaoK related to #1190. 😄

**Comment by brockallen** [comment]
@bgever Many of the things you're asking for (including revocation, introspection, extension grants, etc) are already supported by IdentityServer: http://identityserver.io/

**Comment by PinpointTownes** [comment]
@brockallen don't forget that `OAuthMiddleware` is the OAuth2 client base middleware used by the social providers, not the equivalent of Katana's `OAuthAuthorizationServerMiddleware` in vNext.

**Comment by bgever** [comment]
@brockallen For clarity, with "OAuthMiddleware" I mean the OAuth client (consumer) support rather than OAuth server, which base implementation is done with the [`OAuthMiddleware<TOptions>` class](https://docs.microsoft.com/en-us/aspnet/core/api/microsoft.aspnetcore.authentication.oauth.oauthmiddleware-1).

There are framework and contributed ([aspnet-contrib/AspNet.Security.OAuth.Providers](aspnet-contrib/AspNet.Security.OAuth.Providers) - kudos to @PinpointTownes) implementations/providers for it, or you can even write your own like I did with the Meetup one in my linked example project. Maybe even a generic one for IdentityServer with support for all these fancy features would be nice. 😉 Especially since we're likely going to upgrade our current ASP.NET Identity Framework 2 based OAuth server to IdentityServer4 at some point. 😄

**Comment by brockallen** [comment]
Yes, I misunderstood -- I thought you were asking Microsoft to add those features as a server feature.

**Comment by bgever** [comment]
No worries - I appreciate the input. Maybe it should have been called the OAuth**Client**Middleware in the first place. Naming stuff is hard. 😏

**Comment by PinpointTownes** [comment]
> Maybe even a generic one for IdentityServer with support for all these fancy features would be nice.

Given that IdSrv is an OIDC server, the OIDC client middleware would probably be a better choice :smile: 

> There are framework and contributed (aspnet-contrib/AspNet.Security.OAuth.Providers - kudos to @PinpointTownes) implementations/providers for it, or you can even write your own like I did with the Meetup one in my linked example project.

Haha, thanks! Feel free to submit your provider if you want to see included in the aspnet-contrib stuff :smile:

**Comment by HaoK** [comment]
@bgever we are already overbooked for 2.0 right now, so this isn't something we have immediate plans to look at, so if this is something important to you, a PR might be the best way to get the ball rolling on what you have in mind, plus I believe we got things like SaveTokens in the base OAuth implementation (from a @PinpointTownes PR)

**Comment by bgever** [comment]
@HaoK I was anticipating this as we're so close to the 2.0 release. I don't mind getting started on a PR, as I'll need to make a custom implementation anyhow. It would be great to have this eventually built into the framework, with an additional security review by your team as I'm by no means a security expert.

Should I branch of `dev`? I'm thinking whether: 1) I can expand on `OAuthMiddleware<TOptions>` and then include into my project my own fork, or 2) should I start by creating a copy of the `OAuthMiddleware<TOptions>` class in a separate lib for later replacement of the original? As I'd would want to do some actual integration tests aside from unit tests. My current projects are on 1.1, and would upgrade to 2.0 when available. Would be good to hear your recommended approach on this.

**Comment by bgever** [comment]
> > Maybe even a generic one for IdentityServer with support for all these fancy features would be nice.
>
> Given that IdSrv is an OIDC server, the OIDC client middleware would probably be a better choice 😄

If OpenID Connect (OIDC) can be used in the application, then it could be a better approach than the proposed flows. However, unless I misunderstood, the OIDC client middleware doesn't support the OAuth password or JWT/SAML SSO grant flows. When this issue is closed, the OAuthMiddleware should support these grant flows, also with IdentityServer.

The scenario I'd like to support is where the application (ASP.NET Core site) has its own (custom styled) login form (username & password) or SSO token forwarding endpoint, but handles authentication on the OAuth server. On successful authentication (tokens are returned) then the site maintains a session for the user based on the OAuth response with cookie authentication.

> There are framework and contributed (aspnet-contrib/AspNet.Security.OAuth.Providers - kudos to @PinpointTownes) implementations/providers for it, or you can even write your own like I did with the Meetup one in my linked example project.
>
> Haha, thanks! Feel free to submit your provider if you want to see included in the aspnet-contrib stuff 😄

Yes, I'm planning to submit a PR for the Meetup provider once I've tested it a bit more on what user info to include in the claims. 😄 Currently it only authenticates, but meetup group info isn't included yet.

**Comment by bgever** [comment]
@HaoK Looking at the impact of aspnet/Announcements#232, I don't think I can start working on a PR for 2.0 as we're still dependent on 1.1. However, as I was able to make a custom build of 1.1 and include that in my projects, I can start working on an implementation based on the current 1.1 OAuthMiddleware. Then, once 2.0 is out and our project switches over, I can refactor the code and make a PR for 2.0.

One thing I'm currently stuck on is how to pass the username & password or JWT/SAML token to the OAuthMiddleware. I feel I should use `HttpContext.Authentication.ChallengeAsync` method. That means passing the values in custom fields on the [`AuthenticationProperties`](https://github.com/aspnet/HttpAbstractions/blob/rel/1.1.2/src/Microsoft.AspNetCore.Http.Abstractions/Authentication/AuthenticationProperties.cs). Does that sound right?

Unfortunately the excellent documentation :trollface: on the AuthenticationManager was of no use.

**Comment by bgever** [comment]
I'm giving up on the PR for the 1.1 codebase, as I'm struggling for days to find my way around the poorly documented interfaces, abstractions, overrides and callback logic. The `OAuthHandler`'s current inheritance of `RemoteAuthenticationHandler` is a blocker, which would require complex switching logic to avoid a redirect.

The authentication refactor for 2.0 is much welcome and improves on these parts, however I would suggest to document the code with explanations *why* certain behaviour is in place.

Heck, at a **minimum** consider to document the essential interfaces like [`AuthenticationManager`](https://github.com/aspnet/HttpAbstractions/blob/haok/auth2/src/Microsoft.AspNetCore.Http.Abstractions/Authentication/AuthenticationManager.cs) and its params, even in the 1.1 codebase...

**Comment by Eilon** [comment]
Moving to Backlog because there are no plans to implement this at this time.
